### PR TITLE
Remove enforcing array consistency from CoordinateArraySequence

### PR DESCRIPTION
**Pull request by jodygarnett**
Removes check on array used for CoordinateArraySequence, documenting instead that the array must be consistent with the dimension and measures provided.  Refactored the protected enforceConsistency method out to the Coordintes utility class for any client code that ends up working with mixed coordines.

### CODE REVIEW

File: modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java [hunk 1]
```diff
@@ -168,35 +171,7 @@ public CoordinateArraySequence(CoordinateSequence coordSeq)
...
-            duplicate.setCoordinate(coordinate);
-            copy[i] = duplicate;
-          }
-          else {
-            copy[i] = coordinate;
-          }
-       }
-       return copy;
-     }
+     return Coordinates.enforceConsistency(array, dimension, measures);
```
Comment (davidblasby) [inline modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java:174]: should this method be removed?  (from p2p conversation)

File: modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java [hunk 2]
```diff
@@ -46,7 +48,100 @@ public static Coordinate create(int dimension, int measures)
...
+   * Utility method ensuring array contents are of the specified dimension and measures.
+   * <p>
+   * Array is modified in place if required, coordinates are replaced in the array as required
+   * to ensure all coordinates have the same dimension and measures.
+   * </p>
+   *
+   * @param array
+   * @return array returned, or copy created if required to enforce consistency.
+   */
+  public static Coordinate[] enforceConsistency(Coordinate[] array)
```
Comment (davidblasby) [inline modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:62]: test case?  Or is it covered elsewhere?

File: modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java [hunk 3]
```diff
@@ -46,7 +48,100 @@ public static Coordinate create(int dimension, int measures)
...
+   * Array is returned unmodified if consistent, or a copy of the array is made with
+   * each inconsistent coordinate duplicated into an instance of the correct dimension and measures.
+   * </p></>
+   *
+   * @param array coordinate array
+   * @param dimension
+   * @param measures
+   * @return array returned, or copy created if required to enforce consistency.
+   */
+  public static Coordinate[] enforceConsistency(Coordinate[] array,int dimension, int measures)
```
Comment (davidblasby) [inline modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:112]: test case?  Or is it covered elsewhere?

File: modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java [hunk 4]
```diff
@@ -46,7 +48,100 @@ public static Coordinate create(int dimension, int measures)
...
+   * Array is modified in place if required, coordinates are replaced in the array as required
+   * to ensure all coordinates have the same dimension and measures.
+   * </p>
+   *
+   * @param array
+   * @return array returned, or copy created if required to enforce consistency.
+   */
+  public static Coordinate[] enforceConsistency(Coordinate[] array)
+  {
+    if( array == null ){
```
Comment (davidblasby) [inline modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:64]: if (array == null ) {

File: modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java [hunk 5]
```diff
@@ -46,7 +48,100 @@ public static Coordinate create(int dimension, int measures)
...
+      return array;
+    }
+    // step one check
+    int dimension = 0;
+    int measures = 0;
+    boolean isConsistent=true;
```
Comment (davidblasby) [inline modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:70]: the spacing isn't consistent with above ;)

File: modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java (hunk 1 above)
Comment (jodygarnett) [reply modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java:174]: Thanks you are correct it moved to CoordinateArrays utility class for client code to call if needed

File: modules/core/src/main/java/org/locationtech/jts/geom/LinearRing.java [hunk 6]
```diff
@@ -132,7 +132,7 @@ protected LinearRing copyInternal() {
     return new LinearRing(points.copy(), factory);
   }
 
-  protected LinearRing reverse()
+  public LinearRing reverse()
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/geom/LinearRing.java:135]: Would be better if this was not part of this PR

File: modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java (hunk 2 above)
Comment (jodygarnett) [reply modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:62]: I will add a test case, thanks! Previously it was covered by other tests

File: modules/core/src/main/java/org/locationtech/jts/geom/LinearRing.java (hunk 6 above)
Comment (jodygarnett) [reply modules/core/src/main/java/org/locationtech/jts/geom/LinearRing.java:135]: yes it would, let me fix and then rebase this PR.

File: modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java [hunk 7]
```diff
@@ -11,6 +11,8 @@
  */
 package org.locationtech.jts.geom;
 
+import java.lang.reflect.Array;
```
Comment (davidblasby) [inline modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:14]: probably not used?

**Comment by jodygarnett** [comment]
Thanks for review everyone

File: modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java [hunk 8]
```diff
@@ -75,8 +75,11 @@ public CoordinateArraySequence(Coordinate[] coordinates, int dimension) {
   
   /**
    * Constructs a sequence based on the given array 
-   * of {@link Coordinate}s (the
-   * array is not copied).
+   * of {@link Coordinate}s (the array is not copied).
+   * <p>
+   * Ensure the array is of the indicated dimension and measures. The method
```
Comment (jnh5y) [inline modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java:80]: Hmm... I wonder if this explanation should be expanded.  

The word "Ensure" is throwing me here.  Read quickly, could one think that the method is ensuring consistency?  (Like it was previously?)  

Also, if the requirement is on the inputs, it seems like it'd be better on the description of the coordinates variable.

Honestly, mixing Coordinates and using the common ordinates ought to be fine.  Right?

As a concrete suggestion how about something this:

"CoordinateArray should likely contain consistent implementations of the Coordinate for most uses.  Mixing types may lead to issues.  Callers can create such consistent arrays using {@link CoordinateArrays#enforceConsistency(Coordinate[])}."

**Comment by jodygarnett** [comment]
@jnh5y I have improved the javadoc, David also found a some code improvements to be made. Please merge when the automated checks are passed.

**Comment by jodygarnett** [comment]
Hold up I see Martin marked this as draft, sadly I removed that before noticing he just did so 10 mins ago.

@dr-jts what is up?

**Comment by dr-jts** [comment]
> Hold up I see Martin marked this as draft, sadly I removed that before noticing he just did so 10 mins ago.
> 
> @dr-jts what is up?
Well it appears to still be changing.  But I guess as a result of review.  Is there a way to sign off on review?
